File: backend/src/main/kotlin/es/uib/record/backend/publications/domain/model/PublicationStatus.kt
package es.uib.record.backend.publications.domain.model

enum class PublicationStatus {
    PLANNED,
    SUBMITTED,
    UNDER_REVIEW,
    MINOR_REVISION,
    MAJOR_REVISION,
    REJECTED,
    ACCEPTED,
    PUBLISHED;

    /** Estados a los que se puede transicionar desde este estado. */
    fun allowedTransitions(): Set<PublicationStatus> = ALLOWED_TRANSITIONS[this] ?: emptySet()

    /** Indica si la transición de este estado a [target] está permitida. */
    fun canTransitionTo(target: PublicationStatus): Boolean = target in allowedTransitions()

    /**
     * Indica si es un estado final (terminal) del ciclo de vida: no admite más transiciones. Hoy
     * son REJECTED y PUBLISHED. Se deriva de la máquina de estados para no duplicar la lista.
     */
    fun isFinal(): Boolean = allowedTransitions().isEmpty()

    private companion object {
        /**
         * Máquina de estados del ciclo de vida de una publicación. REJECTED y PUBLISHED son estados
         * terminales para los cambios de estado genéricos ([changeStatus]).
         *
         * El reenvío de una publicación rechazada (REJECTED -> SUBMITTED cambiando de journal) se
         * modela aparte en [Publication.resubmit], que es una operación atómica con sus propias
         * reglas y no abre esta transición genérica.
         */
        val ALLOWED_TRANSITIONS: Map<PublicationStatus, Set<PublicationStatus>> =
            mapOf(
                PLANNED to setOf(SUBMITTED),
                SUBMITTED to setOf(UNDER_REVIEW, REJECTED),
                UNDER_REVIEW to setOf(MINOR_REVISION, MAJOR_REVISION, ACCEPTED, REJECTED),
                MINOR_REVISION to setOf(UNDER_REVIEW),
                MAJOR_REVISION to setOf(UNDER_REVIEW),
                ACCEPTED to setOf(PUBLISHED),
                REJECTED to emptySet(),
                PUBLISHED to emptySet(),
            )
    }
}

File: frontend/src/app/features/publications/publication-status.ts
import { PublicationStatus } from '@core/api';

/**
 * Máquina de estados del ciclo de vida de una publicación. Es el espejo en frontend
 * de la definición de dominio del backend (PublicationStatus.kt). El backend sigue
 * siendo la fuente de verdad y valida toda transición; aquí solo se usa para mostrar
 * únicamente las transiciones válidas en la UI.
 *
 * REJECTED y PUBLISHED son estados terminales.
 */
const ALLOWED_TRANSITIONS: Record<PublicationStatus, PublicationStatus[]> = {
  [PublicationStatus.Planned]: [PublicationStatus.Submitted],
  [PublicationStatus.Submitted]: [PublicationStatus.UnderReview, PublicationStatus.Rejected],
  [PublicationStatus.UnderReview]: [
    PublicationStatus.MinorRevision,
    PublicationStatus.MajorRevision,
    PublicationStatus.Accepted,
    PublicationStatus.Rejected,
  ],
  [PublicationStatus.MinorRevision]: [PublicationStatus.UnderReview],
  [PublicationStatus.MajorRevision]: [PublicationStatus.UnderReview],
  [PublicationStatus.Accepted]: [PublicationStatus.Published],
  [PublicationStatus.Rejected]: [],
  [PublicationStatus.Published]: [],
};

/** Estados a los que se puede transicionar desde [status]. */
export function allowedStatusTransitions(status: PublicationStatus): PublicationStatus[] {
  return ALLOWED_TRANSITIONS[status] ?? [];
}

/** Severidad de PrimeNG (`p-tag`) asociada a cada estado de publicación. */
export function publicationStatusSeverity(
  status: PublicationStatus,
): 'success' | 'info' | 'warn' | 'danger' {
  switch (status) {
    case PublicationStatus.Published:
    case PublicationStatus.Accepted:
      return 'success';
    case PublicationStatus.Rejected:
      return 'danger';
    case PublicationStatus.UnderReview:
    case PublicationStatus.MinorRevision:
    case PublicationStatus.MajorRevision:
      return 'warn';
    default:
      return 'info';
  }
}
